ccvm_simulators/solvers/dl_solver.py

```diff
@@ -158,14 +180,25 @@ def _calculate_grads_boxqp(self, c, s, S=1):
         Args:
             c (torch.Tensor): In-phase amplitudes of the solver
             s (torch.Tensor): Quadrature amplitudes of the solver
+            lower_limit (float): The lower bound of the box constraints. Defaults to 0.
+            upper_limit (float): The upper bound of the box constraints. Defaults to 1.
             S (float): The saturation value of the amplitudes. Defaults to 1.
 
         Returns:
             tuple: The calculated change in the variable amplitudes.
         """
 
-        c_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", c / S + 1, self.q_matrix) / S
-        c_grad_3 = self.v_vector / 2 / S
+        c_grad_1 = (
+            0.25
+            * torch.einsum(
+                "bi,ij -> bj",
+                c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit),
+                self.q_matrix,
+            )
+            * (upper_limit - lower_limit)
+            / S
+        )
+        c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
 
         s_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", s / S + 1, self.q_matrix) / S
         s_grad_3 = self.v_vector / 2 / S
```

Comment:
The `s_grad` variables should also be updated similarly here:
```suggestion
        s_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", s * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit) / S
        s_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2*S)
```

Reply:
Ahh, I missed them! Thanks for noticing.

Reply:
What about the calculate_grads function? It also has these lines but I haven't updated them